### ISSUE TITLE
Syntax highlight/colorize the error messages

### ISSUE THREAD

**Issue by XVilka** [ocaml/ocaml#9328]
Sorry if duplicate - I remember it was discussed somewhere, but found nothing in the opened issues, thus this feature request.

To date many various languages perform a simple syntax highlight or colorize their error messages somehow, to make them more readable. While the text part of the message nowadays is clear enough in most of the cases, adding basic syntax highlight for the reported types or lines of code will be beneficial for a quicker understanding of the problem.

Take the following example:
![ocaml-error](https://user-images.githubusercontent.com/203261/75128528-78b50580-56ff-11ea-8a43-9bc38c33b945.png)

Here the string 
```ocaml
(Ppxlib.Location.t * 'a) list -> (Ppxlib.location.t * 'a) list
```
could be highlighted according to the OCaml syntax
Same with the expected type expression.
Also in the string 
```
Type Ppxlib.Location.t = Warnings.loc is not compatible with type int
```
both `Ppxlib.Location.t = Warnings.loc` and `int` can be of a different color from the rest of the message.
Implementing this will allow to understand the error from the first glance.

**Comment by gasche** [comment]
TL/DR: people are working on this but it's in fact a lot of work, and it has been progressing surely but slowly.

There was a first attempt to do this by @Octachron in https://github.com/ocaml/ocaml/pull/1481 , using what is morally a syntactic diff on the printed representation of the types. It was a lot of work but it end up not-merged, because there were too many positives (differences being highlighted that amounted to different expansion choices, not to real differences). To do this work properly we need a more abstract representation of type errors, and building this representation requires a lot of work, for each different kind of error.

Various contributors have been slowly working on improving error representations, in particular @MekhrubonT in #8828 on type mismatches, and @Octachron and @Drup in ongoing work on module signature mismatches -- their work integrates colors nicely in the output.

**Comment by lpw25** [comment]
@Octachron / @Drup is your ongoing work in the same area as #8828? Because I have some further PRs from that internship that just need finishing off. The current state is [here](http://github.com/lpw25/ocaml/commits/better-error-types) and [here](http://github.com/lpw25/ocaml/commits/use-better-error-types). They're probably not in a state for others to work on them, but you should be able to see if we are duplicating effort.

**Comment by Octachron** [comment]
If I am reading correctly, the two should be really complimentary: our [work](https://github.com/Octachron/ocaml/commits/semdiff_includemod) is more concerned with reorganizing the global flow of information in includecore. The idea is to make the full error tree available for the error printer. The initial PR (hopefully coming sometime this week) will essentially target functor application and inclusion, and a bit of clarification for module type equality, while keeping otherwise the same message structure. If I understand correctly, your PRs are essentially refining the leaves (and nodes?) of this tree wish should mesh well with this work. Even if the merge conflict might be a bit painful.

**Comment by lpw25** [comment]
> If I understand correctly, your PRs are essentially refining the leaves (and nodes?) of this tree

That's right.

> should mesh well with this work

Yeah, it sounds like they should compliment each other nicely.

**Comment by Drup** [comment]
The 2 problems are slightly orthogonal. The work we have started is about structured error messages which allow for better explanations. Code highlighting should be implemented simply by tagging (using `Format` tags) the output, which is very easy to do, but also quite labor-intensive. That could be started at any point, rather independently.

**Comment by XVilka** [comment]
Not really stale.

**Comment by XVilka** [comment]
It's notable that aforementioned semantic diffing for ill-typed functors patch was merged in the master: https://github.com/ocaml/ocaml/commit/d4dd566e3ee5c5fc2ed1b328ccdcf347804cbd66 that is also uses some colors 👍 : 

![image](https://user-images.githubusercontent.com/203261/113967937-b1149600-9864-11eb-8b86-5c074db33ace.png)

**Comment by gasche** [comment]
What is the meaning of the color schemes used there? If I understand correctly, green is used when an argument match the expected type (of its corresponding formal parameter), magenta is used for both the argument and the formal parameter when there is a mismatch, and red is used when an argument (or formal parameter) has no corresponding formal parameter (or argument).

In the present PR, we don't have a case where a type matches its expectation (where they would both be green); if we reused the colors of the example example, I guess the two types would be displayed in magenta: we have exactly one expectation and one provided type, and they don't match.

**Comment by XVilka** [comment]
There is a label "error-messages", I suggest to assign it to this issue, it will finding all these error message improvements easier:
https://github.com/ocaml/ocaml/labels/error-messages

**Comment by Octachron** [comment]
I agree that some level of syntactic highlighting will help with the readability of error messages. However, colors are a scarce resources and I am not convinced that syntactic category are the place to spend this resource: I would rather focus on highlighting meaningful differences.
